mysql-test/mysql-test-run.pl

```diff
@@ -929,7 +929,8 @@ ($$$)
 
                 rename $log_file_name, $log_file_name.".failed";
               }
-	      delete($result->{result});
+            $result->{retries}= $retries;
+            push(@$completed, $result);
```

Comment:
This looks a bit fishy. Apparently the old code logic was to sent serialized `$result` to the `$sock`, but without `$result->{result}` and with adjusted `$result->{retries}`.

Now you
* send `$result` **with** `$result->{result}` and
* you store `$result` in the `@$completed` array **with the adjusted** `$result->{retries}`.
May be that's fine, but both look rather suspicious.

So, either, please, verify that it's ok not to delete `$result->{result}` when sending and that you don't care what the value of `$result->{retries}` is in the array. Or preserve them, for example, with
```perl
  {
    local @$result->{retries, result};
    delete $result->{result};
    $result->{retries}= $retries+1;
    $result->write_test($sock, 'TESTCASE');
  }
  push(@$completed, $result);
```
this way it'll send exactly the same data as before, but will store the unmodified `$result` in the array

Reply:
If we delete the values of `$result->{result}`, we will lose the values of these results. XML example:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="" failures="5" name="" tests="6" time="0.003">
        <testsuite disabled="" errors="" failures="1" hostname="mdbci-f3ellluf-1617800556-node" id="0" name="main" package="" skipped="" tests="6" time="0.003" timestamp="2021-04-08T07:21:22Z">
                <testcase assertions="" classname="main" name="t1" status="MTR_RES_PASSED" time="0.001" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="" time="0.000" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="" time="0.000" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="" time="0.000" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="" time="0.001" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="MTR_RES_FAILED" time="0.001" combinations="">
                        <failure message="" type="MTR_RES_FAILED">
<![CDATA[CURRENT_TEST: main.t2
mysqltest: At line 1: query 'select null into outfile 'f'' failed: 1086: File 'f' already exists
]]>
                        </failure>
                </testcase>
        </testsuite>
</testsuites>
```
As for the effect of saving retries, they are now only used to output an [additional substring](https://github.com/MariaDB/server/blob/10.2/mysql-test/lib/mtr_report.pm#L169), that is, the existence or non-existence of this value is used.

And the code you provided is not interpreted.

Reply:
About the fact that this patch works correctly and does not crash. I conducted a large number of combinations of falls and non-falls of the test, including those described above, always the result was correct. XML example:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="" failures="5" name="" tests="6" time="0.002">
        <testsuite disabled="" errors="" failures="3" hostname="mdbci-f3ellluf-1617800556-node" id="0" name="main" package="" skipped="" tests="6" time="0.002" timestamp="2021-04-08T09:40:16Z">
                <testcase assertions="" classname="main" name="t1" status="MTR_RES_PASSED" time="0.000" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="MTR_RES_FAILED" time="0.000" combinations="">
                        <failure message="" type="MTR_RES_FAILED">
<![CDATA[CURRENT_TEST: main.t2
mysqltest: At line 1: query 'select null into outfile 'f'' failed: 1086: File 'f' already exists
]]>
                        </failure>
                </testcase>
                <testcase assertions="" classname="main" name="t2" status="MTR_RES_PASSED" time="0.001" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="MTR_RES_FAILED" time="0.001" combinations="">
                        <failure message="" type="MTR_RES_FAILED">
<![CDATA[CURRENT_TEST: main.t2
mysqltest: At line 1: query 'select null into outfile 'f'' failed: 1086: File 'f' already exists
]]>
                        </failure>
                </testcase>
                <testcase assertions="" classname="main" name="t2" status="MTR_RES_PASSED" time="0.000" combinations="" />
                <testcase assertions="" classname="main" name="t2" status="MTR_RES_FAILED" time="0.000" combinations="">
                        <failure message="" type="MTR_RES_FAILED">
<![CDATA[CURRENT_TEST: main.t2
mysqltest: At line 1: query 'select null into outfile 'f'' failed: 1086: File 'f' already exists
]]>
                        </failure>
                </testcase>
        </testsuite>
</testsuites>
```

Reply:
1. about result — my concern is that the old code intentionally removed result before sending a test to a worker. You changed that. I asked to verify that it's fine. **Of course** XML would be right, workers don't create XML, what you changed (or, potentially, broke) does not affect XML generation.
2. the bit of code that I specified was a pseudo-perl example, showing the idea of a fix, it's normal if it isn't a valid perl. But okay, here you are, a working example:
```perl
              {
                local @$result{'retries', 'result'};
                delete $result->{result};
                $result->{retries}= $retries+1;
                $result->write_test($sock, 'TESTCASE');
              }
              push(@$completed, $result);
```
3. retries — yes, that answers my question, thanks.

Reply:
Thank you for the working example! Fixed saving `$result->{result}`.